shiny/_template_utils.py

```diff
@@ -104,34 +111,177 @@ def download_and_extract_zip(url: str, temp_dir: Path):
+            except Exception as err:
+                errors.append(str(err))
+                pass
+
+        if extracted_dir is None:
+            raise click.ClickException(
+                f"Failed to download repository from GitHub {cli_url(github)}. "
+                + "Please check the URL or GitHub spec and try again.\n"
+                + "We received the following errors:\n"
+                + textwrap.indent("\n".join(errors), "  ")
+            )
+
+        template_dir = extracted_dir / spec.path
 
-        template_dir = os.path.join(
-            temp_dir, f"{repo_name}-{branch_name}", subdirectory
+        if not os.path.exists(template_dir):
+            raise click.ClickException(
+                f"Template directory '{cli_input(spec.path)}' does not exist in {cli_field(spec_cli)}."
+            )
+
+        return app_template_questions(
+            template=template,
+            mode=mode,
+            template_dir=Path(template_dir),
+            dest_dir=dest_dir,
         )
 
-        if not os.path.exists(template_dir):
-            raise Exception(f"Template directory '{template_dir}' does not exist")
 
-        directory = repo_name + "-" + branch_name
-        path = temp_dir / directory / subdirectory
-        return app_template_questions(mode=mode, template_dir=path, dest_dir=dest_dir)
+def github_zip_url(spec: GithubRepoLocation) -> Generator[str]:
+    for suffix in ["refs/heads/", "refs/tags/", ""]:
+        url = f"https://github.com/{spec.repo_owner}/{spec.repo_name}/archive/{suffix}{spec.ref}.zip"
+        yield url
+
+
+@dataclass
+class GithubRepoLocation:
+    repo_owner: str
+    repo_name: str
+    ref: str
+    path: str
+
+
+def parse_github_arg(x: str) -> GithubRepoLocation:
+    if re.match(r"^(https?:)//|github[.]com", x):
+        return parse_github_url(x)
+    return parse_github_spec(x)
+
+
+def parse_github_spec(spec: str):
+    """
+    Parses GitHub repo + path spec in the form of:
+
+    * {repo_owner}/{repo_name}@{ref}:{path}
+    * {repo_owner}/{repo_name}:{path}@{ref}
+    * {repo_owner}/{repo_name}:{path}
+    * {repo_owner}/{repo_name}/{path}
+    * {repo_owner}/{repo_name}/{path}@{ref}
+    * {repo_owner}/{repo_name}/{path}?ref={ref}
+    """
+
+    # We split the spec into parts, using a capture group so that the splitting
+    # characters are retained in the path parts. Then we know that {repo_owner} /
+    # {repo_name} come first in the first three parts, reducing the problem to parsing
+    # ref and path in the remaining parts.
+    parts = re.split(r"(:|@|[?]ref=|/)", spec)
+
+    if len(parts) < 3:
+        raise click.BadParameter(
+            f"Could not parse as GitHub spec: '{cli_input(spec)}'"
+            + ". Please use the format "
+            + cli_field("{repo_owner}/{repo_name}@{ref}:{path}")
+            + ".",
+            param_hint="--github",
+        )
+
+    repo_owner, _, repo_name = parts[:3]
+    path = ""
+    ref = ""
+
+    which = "path"
```

Comment:
Seems like you could get rid of the `which`?

Reply:
That for loop has two parts: 

1. if we're looking at a separator, we decide what the separator means
2. if it's not a separator, decide where to add the part we're considering

So I can't get rid of `which="path"`, aka the initial assumption, but we could move it other places, but I'm not seeing a scenario that I find to be more clear. What do you have in mind?

Reply:
Ohhh, duh, sorry for the noise.